compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIESimplifyDmaWaits.cpp

```diff
@@ -0,0 +1,158 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree-amd-aie/IR/AMDAIEOps.h"
+#include "iree-amd-aie/Transforms/AMDAIEDmaUtils.h"
+#include "iree-amd-aie/Transforms/AMDAIEUtils.h"
+#include "iree-amd-aie/Transforms/Passes.h"
+#include "iree-amd-aie/aie_runtime/iree_aie_runtime.h"
+#include "mlir/IR/Iterators.h"
+#define DEBUG_TYPE "iree-amdaie-simplify-dma-waits"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+namespace {
+
+/// Traverses the control code in reverse, ensuring that for each connection,
+/// only one DMA wait op is retained for every maximum queue size.
+LogicalResult simplifyDmaWaits(AMDAIE::AMDAIEDeviceModel deviceModel,
+                               AMDAIE::WorkgroupOp workgroupOp) {
```

Comment:
This is a nit, but you could take a `AMDAIE::ControlCodeOp` as parameter instead as the workgroup is only needed to retrieve the control code op, so this would simplify the function a bit and make in more easily reusable.